[Spring 지하철 노선도 - 3단계] 애쉬(김동호) 미션 제출합니다.

dongho108 opened the pull request:
안녕하세요 핀!미션이 어려워서 제출이 좀 늦었습니다.기능구현은 다 완료해서 일단 PR 날리겠습니다!

피드백 주실때까지 리팩토링 하고 있겠습니다!

dongho108 committed:
refactor: findById시 Optional 사용하지 않도록 수정

dongho108 committed:
refactor: exception의 매개변수가 하나이고 명시되어 있으므로 클래스명 생략

dongho108 committed:
docs: 추가된 기능목록 작성

dongho108 committed:
feat: 노선 추가시 추가정보 입력받고 저장하는 기능 추가

  - upStationId: 상행 종점
  - downStationId: 하행 종점
  - distance: 두 종점간의 거리

dongho108 committed:
refactor: query의 테이블 명을 schema와 동일하게 수정

dongho108 committed:
refactor: StationResponse Station을 인자로 받는 정적팩토리 메서드 추가

dongho108 committed:
refactor: 조회 결과가 없을 때의 예외는 not found를 반환하도록 수정

dongho108 committed:
refactor: update시 조회한 객체를 바탕으로 수행하도록 수정

- 조회 결과가 없을때 Not Found 예외를 반환하도록 수정

dongho108 committed:
refactor: LineDao에서 findById시에 Optional을 반환하도록 수정

dongho108 committed:
refactor: spring bean validation 적용

dongho108 committed:
refactor: LineRequest에 validation 적용

dongho108 committed:
refactor: StationRequest에 validation 적용

dongho108 committed:
refactor: Line api e2e 테스트 코드 리팩토링

dongho108 committed:
refactor: Station api e2e 테스트 코드 리팩토링

dongho108 committed:
refactor: request mapping에 consumes, produces 옵션 제거

dongho108 committed:
refactor: Section 도메인이 갖는 필드 수정

- line id x
- upStationId x upStation o
- downStationId x downStation o

dongho108 committed:
refactor: Line이 List<Section>을 갖도록 수정

- findById, findAll시에 SECTION과 STATION을 조인해 만들어줌
- LineResponse에는 List<Station>을 담아 보내줌

dongho108 committed:
feat: 노선에 구간을 삽입하는 기능 구현

dongho108 committed:
feat: 노선에 구간을 삽입하는 api 등록

dongho108 committed:
feat: 노선의 구간을 삭제하는 기능 구현

dongho108 committed:
feat: 노선의 구간을 삭제하는 api 등록

dongho108 committed:
refactor: 메서드 분리

dongho108 committed:
refactor: 메서드 분리

dongho108 committed:
refactor: 메서드 분리

dongho108 committed:
refactor: 메서드 순서 변경

dongho108 committed:
refactor: Line에 Line과 Sections를 받는 정적팩토리 메서드 구현

dongho108 committed:
refactor: 필요없는 정적팩토리 메서드 제거

dongho108 committed:
style: 필요없는 괄호 제거

lalize commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -15,6 +19,33 @@ public Line(Long id, String name, String color) {
+    public Line(Long id, String name, String color, List<Section> sections) {
+        this(id, name, color);
+        this.sections = Sections.of(sections);
+    }
```
모든 필드를 한번에 초기화하는 주생성자를 만들고 호출하면 어떨까요?

lalize replied:
이거 확인 부탁드릴게요!

dongho108 replied:
```java
    private Line(Long id, String name, String color, Sections sections) {
        this(id, name, color);
        this.sections = sections;
    }
```

`주생성자`를 만들라는 말씀이 이해가 안돼서 이렇게 주생성자를 호출하는 부생성자를 만들어 피드백 반영을 했는데 이것을 말씀하신 것이 맞을까요?

dongho108 replied:
지금은
```java
private Line(Long id, String name, String color, Sections sections) {
        this.id = id;
        this.name = name;
        this.color = color;
        this.sections = sections;
    }

    public Line(String name, String color) {
        this(null, name, color, null);
    }

    public Line(Long id, String name, String color) {
        this(id, name, color, null);
    }
```

이렇게 주생성자에 null을 주는식으로 고쳤습니다!

lalize commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -1,37 +1,57 @@
-        return LineResponse.from(savedId, line);
+        Long savedLineId = lineDao.save(line);
+
+        Station upStation = stationDao.findById(lineRequest.getUpStationId())
+            .orElseThrow((throwEmptyStationException()));
+        Station downStation = stationDao.findById(lineRequest.getDownStationId())
+            .orElseThrow(throwEmptyStationException());
+
+        sectionDao.save(Section.from(upStation, downStation, lineRequest.getDistance()), savedLineId);
```
상행과 하행이 같아도 등록이 되고 노선 목록 조회가 정상적으로 되지 않네요.

dongho108 replied:
상행과 하행이 같아도 등록이 된다는게
1. 구간이 상행, 하행이 같은 것을 말씀하시는 건가요?
2. 아니면 1호선이 강남역 -> 선릉역을 가지고, 2호선이 또 강남역 -> 선릉역을 가지는 상황을 말씀하시는 건가요?

1번의 경우는 Section 생성자에서 검증하도록 추가해주었습니다.
2번의 경우는 충분히 가능한 상황이라고 보고 있습니다!

그리고 노선 목록 조회가 어느 경우에서 정상적으로 안되는지 모르겠습니다..ㅠ

lalize commented on src/main/java/wooteco/subway/domain/Line.java:
```diff
@@ -1,9 +1,13 @@
-    private final String color;
+    private String name;
+    private String color;
+    private Sections sections;
```
Line이 구간을 가지도록 하셨네요 👍

lalize commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,195 @@
+        LinkedList<Section> flexibleSections = new LinkedList<>(this.sections);
+        for (int i = 0; i < flexibleSections.size(); i++) {
+            Section sectionInLine = flexibleSections.get(i);
+            if (insertSection(section, flexibleSections, i, sectionInLine)) {
```
depth가 2네요~

dongho108 replied:
IntStream과 findFirst를 이용해 depth를 줄였습니다.

근데 지금 filter에서 데이터수정 로직까지 수행하는 것 같아 수정해보고 있습니다.

lalize commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -41,6 +61,44 @@ public boolean deleteById(Long id) {
+        return lineDao.updateById(id, line);
+    }
+
+    public void insertSection(Long id, SectionRequest sectionRequest) {
+        Line line = lineDao.findById(id)
+            .orElseThrow(throwEmptyLineResultException());
+        Station upStation = stationDao.findById(sectionRequest.getUpStationId())
+            .orElseThrow(throwEmptyStationException());
+        Station downStation = stationDao.findById(sectionRequest.getDownStationId())
+            .orElseThrow(throwEmptyStationException());
+
+        Section section = Section.from(upStation, downStation, sectionRequest.getDistance());
+        Optional<Section> updatedSection = line.insertSection(section);
+        updatedSection.ifPresent(sectionDao::update);
+        sectionDao.save(section, line.getId());
+    }
+
+    public void deleteStation(Long lineId, Long stationId) {
+        Station station = stationDao.findById(stationId)
+            .orElseThrow(throwEmptyStationException());
+        Line line = lineDao.findById(lineId)
+            .orElseThrow(throwEmptyLineResultException());
+        UpdatedSection updatedSection = line.deleteStation(station);
+        if (updatedSection.hasUpdatedSection()) {
+            sectionDao.update(updatedSection.getUpdatedSection());
+        }
+        sectionDao.delete(updatedSection.getDeletedSectionId());
+    }
```
구간을 추가하거나 삭제할 때 영향 받는 기존 구간만 **DB에 반영하기 위해서** 도메인 객체가 설계된 부분이 있는 것 같은데요~ 어떤 생각을 가지고 설계를 했는지 궁금해요.

dongho108 replied:
이 부분에 대해서 잘못날린 PR에서 질문했었는데요!
최대한 도메인이 DAO를 모르는 관점에서 짜보려고 했는데도 도메인 객체가 DB를 알고있는 코드가 나와버렸네요

그래서 지금은
insert의 반환값을 성공했는지 여부,
delete의 반환값을 삭제된 sectionId를 반환하게 수정했습니다!

lalize commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -1,37 +1,57 @@
-        return LineResponse.from(savedId, line);
+        Long savedLineId = lineDao.save(line);
+
+        Station upStation = stationDao.findById(lineRequest.getUpStationId())
+            .orElseThrow((throwEmptyStationException()));
+        Station downStation = stationDao.findById(lineRequest.getDownStationId())
+            .orElseThrow(throwEmptyStationException());
```
`findStationById` private 메서드를 만드는 건 어떨까요?

dongho108 replied:
너무좋아요!

lalize commented on src/main/java/wooteco/subway/ui/LineController.java:
```diff
@@ -28,19 +30,19 @@ public LineController(LineService lineService) {
 
-    @PostMapping(consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
-    public ResponseEntity<LineResponse> createLine(@RequestBody LineRequest lineRequest) {
+    @PostMapping()
```
```suggestion
    @PostMapping
```

lalize commented on src/test/java/wooteco/subway/domain/SectionsTest.java:
```diff
@@ -0,0 +1,181 @@
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+class SectionsTest {
```
정상 케이스뿐만 아니라 예외 케이스에 대해서도 테스트해주세요 :)

dongho108 replied:
구간추가할때 몇개의 예외사항을 더 추가했습니다.

또한 구간삭제의 예외테스트가 없어서 추가했습니다!

lalize reviewed (changes requested):
안녕하세요! 애쉬~
3단계 구현 잘 해주셨네요 👍 
몇가지 리뷰 남겼으니 확인 부탁드릴게요~

dongho108 committed:
refactor: 중복되는 메서드 분리

dongho108 committed:
fix: 구간 내의 상행선과 하행선이 같은 경우에도 구간이 생성되는 문제 해결

dongho108 committed:
fix: 상행역과 하행역이 이미 노선에 모두 등록되어 있어도 구간이 추가되던 문제 해결

dongho108 committed:
test: 마지막 구간 제거 오류검증 테스트 추가

dongho108 committed:
refactor: 구간추가나 삭제시 도메인이 DB에 의존하던 설계를 수정

dongho108 committed:
refactor: depth 최대 2 -> 1로 수정

- fori 반복문을 돌며 첫번째로 조건이 맞을때 리턴했었던 코드를
- IntStream.range(..).filter(..).findFirst()로 해결
- 상행선과 하행선이 같을 경우는 예외를 반환
- 이외에 단지 추가가 안됐을 경우에는 false 반환

dongho108 committed:
refactor: 브랜치 병합

dongho108 committed:
style: 쓸모없는 주석 제거

dongho108 committed:
refactor: 합칠 수 있는 라인 합치기

dongho108 committed:
refactor: stream을 불변을 보장하도록 리팩터링

lalize commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -41,6 +57,59 @@ public boolean deleteById(Long id) {
+        Station downStation = findStationById(sectionRequest.getDownStationId());
+        Section section = new Section(upStation, downStation, sectionRequest.getDistance());
+
+        checkInsertSuccess(line.insertSection(section));
```
line.insertSection이 실패하면 예외를 던지면 어떨까요?

dongho108 replied:
좋습니다!!

lalize commented on src/main/java/wooteco/subway/service/LineService.java:
```diff
@@ -1,37 +1,53 @@
 
     public LineResponse save(LineRequest lineRequest) {
         Line line = new Line(lineRequest.getName(), lineRequest.getColor());
-        Long savedId = lineDao.save(line);
-        return LineResponse.from(savedId, line);
+        Long savedLineId = lineDao.save(line);
+
+        Station upStation = findStationById(lineRequest.getUpStationId());
+        Station downStation = findStationById(lineRequest.getDownStationId());
+
+        sectionDao.save(new Section(upStation, downStation, lineRequest.getDistance()), savedLineId);
+        return LineResponse.from(savedLineId, line);
     }
```
Line과 Section은 둘 다 저장이 되거나 둘 다 저장이 되지 않아야 하는데요!
Section 저장이 실패해도 Line은 저장되네요!

테스트로 상행 종점과 하행 종점이 같은 노선을 생성해보세요.
Section 생성 시 validation에 의해 예외가 발생하여 Line만 저장되고 Section은 저장되지 않습니다.

DB 트랜잭션과 `@Transactional`에 대해 학습해보시고 하나의 트랜잭션으로 묶여야 하는 로직들을 확인하여 적용해주세요.

lalize replied:
>상행과 하행이 같아도 등록이 된다는게
-구간이 상행, 하행이 같은 것을 말씀하시는 건가요?
-아니면 1호선이 강남역 -> 선릉역을 가지고, 2호선이 또 강남역 -> 선릉역을 가지는 상황을 말씀하시는 건가요?
1번의 경우는 Section 생성자에서 검증하도록 추가해주었습니다.
2번의 경우는 충분히 가능한 상황이라고 보고 있습니다!
그리고 노선 목록 조회가 어느 경우에서 정상적으로 안되는지 모르겠습니다..ㅠ

노선을 등록할 때 상행 종점과 하행 종점이 같은 경우를 말씀드린거구요~ 그렇게 등록하고나서 노선 목록 조회 오류가 발생했었어요.

dongho108 replied:
`@Transactional`
* `@Transactional` 이 붙은 메서드는 포함하고 있는 작업 중에 하나라도 실패할 경우 전체 작업을 롤백한다.
* `readOnly = true` 속성은 데이터를 수정하지 않는다면(읽기작업수행) 성능이 향상된다. (트랜잭션에 대한 최적화)

상행종점과 하행종점이 같은 노선을 생성할때 Line을 먼저 생성하고 그 line_id로 section을 생성하는데 section을 생성할때 실패하게되면 line만 저장된 상태가 된다는 것을 이해했습니다.
그래서 언급해주신 `@Transactional` 어노테이션을 이용하여 작업중에 하나라도 실패하면 실행했던 작업들이 롤백되도록 수정했습니다!!

lalize commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,209 @@
+    }
+
+    private boolean canInsertSection(Section section, Section sectionInLine) {
+        if (canInsertUpStation(section, sectionInLine) && !canInsertDownStation(section, sectionInLine)) {
+            return true;
+        }
+        if (canInsertDownStation(section, sectionInLine) && !canInsertUpStation(section, sectionInLine)) {
+            return true;
+        }
+        return false;
```
if 문 없이 불 표현식을 바로 반환하면 어떨까요?

lalize replied:
`canInsertUpStation(section, sectionInLine) && !canInsertDownStation(section, sectionInLine)`
`canInsertDownStation(section, sectionInLine) && !canInsertUpStation(section, sectionInLine)`

insertSection에서도 구현을 하고 있는데요.
각각을 메서드로 추출하여 의미를 부여하고 재사용하면 어떨까요?

dongho108 replied:
좋아요!!!

lalize commented on src/main/java/wooteco/subway/domain/Sections.java:
```diff
@@ -0,0 +1,209 @@
+    }
+
+    private void checkContainsStation(Section section) {
+        List<Station> stations = getStations();
+        if (stations.contains(section.getUpStation()) && stations.contains(section.getDownStation())) {
+            throw new IllegalArgumentException("이미 존재하는 상행선과 하행선은 구간에 추가할 수 없습니다.");
+        }
```
sections.contains(section)으로도 이미 존재하는지 확인할 수 있지 않나요?

dongho108 replied:
2 → 4 → 6 이렇게 구간이 있을때 2 → 6을 구간으로 등록하는것도 예외이기 때문에 sections.contains로는 검증이 안될거 같습니다!

lalize reviewed (changes requested):
애쉬 안녕하세요!
몇가지 리뷰 남겨두었으니 확인하여 반영 부탁드릴게요.
pr은 내일 오전에 머지하도록 하겠습니다!

dongho108 committed:
refactor: 이미 했던 검사를 또 하던 코드 수정

dongho108 committed:
refactor: 메서드 분리

dongho108 committed:
refactor: insert가 성공여부를 반환하는 것이 아니라 예외를 던지도록 수정

dongho108 committed:
refactor: `@Transactional`을 이용해 작업 중간에 실패해도 실행한 작업이 롤백되도록 수정

dongho108 committed:
refactor: 모든 필드를 초기화하는 생성자를 사용하도록 수정

dongho108 committed:
refactor: 모든 필드를 초기화하는 주생성자를 사용하도록 수정

dongho108 committed:
remove: 사용하지 않는 객체 삭제

lalize reviewed (approved):
안녕하세요 애쉬~
리뷰 반영 잘해주셨어요 👍 
다음 미션 진행해주세요~
고생하셨습니다~
